cmd/verify.rb

```diff
@@ -1,79 +1,143 @@
+                Utils::Bottles::Tag.from_symbol(bottle_tag)
+              else
+                Utils::Bottles::Tag.new(system: os, arch:)
+              end
 
-          bottle = formula.bottle_for_tag(bottle_tag)
+              bottle = formula.bottle_for_tag(bottle_tag)
 
-          if bottle.nil?
-            opoo "Bottle for tag #{bottle_tag.to_sym.inspect} is unavailable."
-            next
+              if bottle.nil?
+                opoo "Bottle for tag #{bottle_tag.to_sym.inspect} is unavailable."
+                next
+              end
+              bottle.fetch
+              result = check_attestation(bottle, "Homebrew/homebrew-core")
+              if result[:verified]
+                if args.json?
+                  json_results.push(result[:data])
+                else
+                  ohai "Verified #{bottle.name} with tag #{bottle_tag}."
+                end
+              else
+                case result[:error]
+                when JSON::ParserError
+                  opoo "#{bottle.name} with tag #{bottle_tag} returned invalid json: #{result[:message]}"
+                  next
+                when COMMAND_ERROR
+                  opoo "#{bottle.name} with tag #{bottle_tag} unverified, checking backfill signature."
+                  backup_result = check_attestation(bottle, "trailofbits/homebrew-brew-verify")
+                  if backup_result[:verified]
+                    timestamp = backup_result.dig(:data, 0, "verificationResult", "verifiedTimestamps", 0,
+                                                  "timestamp")
+                    if timestamp.nil?
+                      opoo "Unable to verify #{bottle.name} with tag #{bottle_tag} " \
+                           "because json does not have signature timestamp in expected location."
+                      next
+                    end
+                    parsed_timestamp = DateTime.parse(timestamp)
+                    # Backfilled signatures all occured before this date so if we find
+                    # a backfilled signature after this date, we should disregard it.
+                    last_signature_date = DateTime.new(2024, 3, 14)
+                    if parsed_timestamp < last_signature_date
+                      if args.json?
+                        json_results.push(backup_result[:data])
+                      else
+                        ohai "Verified #{bottle.name} with tag #{bottle_tag} using backfill signature."
+                      end
+                    else
+                      opoo "Unable to verify #{bottle.name} with tag #{bottle_tag}. " \
+                           "Backfilled signature dated after last backfill date."
+                    end
+                  end
+                end
+              end
+            end
           end
+        end
+        return unless args.json?
+
+        puts json_results.to_json
+      end
 
-          bottle.fetch
+      def check_attestation(bottle, signing_repo)
```

Comment:
We can do this in a follow-on PR, but: IMO we should put this in a separate `lib/attestation.rb` file, to make it easier for us to upstream/reuse across both `brew verify` and `brew install`.

Reply:
Yeah, that makes sense. Should be quick, but it would be good to get this PR on main so the command works again.